src/communication/response_broker.hpp

```diff
@@ -82,76 +105,112 @@ namespace cubcomm
 
 namespace cubcomm
 {
-  template <typename T_PAYLOAD>
-  response_broker<T_PAYLOAD>::response_broker (size_t a_bucket_count)
+  template <typename T_PAYLOAD, typename T_ERROR>
+  response_broker<T_PAYLOAD, T_ERROR>::response_broker (size_t a_bucket_count, T_ERROR a_no_error)
+    : m_buckets { a_bucket_count, { a_no_error } }
   {
     assert (a_bucket_count > 0);
-    m_buckets.resize (a_bucket_count);
   }
 
-  template <typename T_PAYLOAD>
-  typename response_broker<T_PAYLOAD>::bucket &
-  response_broker<T_PAYLOAD>::get_bucket (response_sequence_number rsn)
+  template <typename T_PAYLOAD, typename T_ERROR>
+  typename response_broker<T_PAYLOAD, T_ERROR>::bucket &
+  response_broker<T_PAYLOAD, T_ERROR>::get_bucket (response_sequence_number rsn)
   {
     return m_buckets[rsn % m_buckets.size ()];
   }
 
-  template <typename T_PAYLOAD>
+  template <typename T_PAYLOAD, typename T_ERROR>
   void
-  response_broker<T_PAYLOAD>::register_response (response_sequence_number a_rsn, T_PAYLOAD &&a_payload)
+  response_broker<T_PAYLOAD, T_ERROR>::register_response (response_sequence_number a_rsn, T_PAYLOAD &&a_payload)
   {
     get_bucket (a_rsn).register_response (a_rsn, std::move (a_payload));
   }
 
-  template <typename T_PAYLOAD>
+  template <typename T_PAYLOAD, typename T_ERROR>
+  void
+  response_broker<T_PAYLOAD, T_ERROR>::register_error (response_sequence_number a_rsn, T_ERROR &&a_error)
+  {
+    get_bucket (a_rsn).register_error (a_rsn, std::move (a_error));
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::bucket (T_ERROR a_no_error)
+    : m_no_error { a_no_error }
+  {
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::~bucket ()
+  {
+    // NOTE: might not hold in the event that a peer server crashes before consuming a response;
+    // in which case the response will linger on in the queue indefinitely
+    assert (m_response_payloads.empty ());
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
   void
-  response_broker<T_PAYLOAD>::bucket::register_response (response_sequence_number a_rsn, T_PAYLOAD &&a_payload)
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::register_response (response_sequence_number a_rsn, T_PAYLOAD &&a_payload)
   {
     {
       std::lock_guard<std::mutex> lk_guard (m_mutex);
 
-      T_PAYLOAD &ent = m_response_payloads[a_rsn];
-      ent = std::move (a_payload);
+      payload_or_error_type &ent = m_response_payloads[a_rsn];
+      ent.m_payload = std::move (a_payload);
+      ent.m_error = m_no_error;
+    }
+    m_condvar.notify_all (); // all because there is more than one thread waiting for data
+  }
+
+  template <typename T_PAYLOAD, typename T_ERROR>
+  void
+  response_broker<T_PAYLOAD, T_ERROR>::bucket::register_error (response_sequence_number a_rsn, T_ERROR &&a_error)
+  {
+    {
+      std::lock_guard<std::mutex> lockg (m_mutex);
```

Comment:
m_send_mutex is already acquired. This may be source of deadlocks. At least add a comment, please.

Reply:
I do not understand the comment.
Either `register_error` or `register_response` is indeed called from the underlying sending thread - the one in `request_queue_autosend::start_thread` via `request_queue_autosend::loop_send_requests` and then `wait_not_empty_and_send_all` and then `send_queue`.
During this time, the client thread - maybe a transaction thread or a system thread - waits in `request_sync_client_server::send_recv` via `response_broker::bucket::get_response` but the wait is guarded by `m_mutex` and `m_condvar` so there should be no problem.